File: php/src/RadixCodec.php
<?php
namespace Vanni\Idmix;

/**
 * 自定义进制文本层：二进制块 ↔ 字符串。
 */
class RadixCodec
{
    private int $base;
    /** @var string[] */
    private array $chars;
    /** @var array<string, int> */
    private array $fromCustom;

    public function __construct(string $alphabet)
    {
        $runes = preg_split('//u', $alphabet, -1, PREG_SPLIT_NO_EMPTY);
        if (count($runes) < 2) {
            throw new \InvalidArgumentException('alphabet must have at least 2 unique characters');
        }
        $this->base = count($runes);
        $this->chars = $runes;
        $this->fromCustom = [];
        foreach ($runes as $i => $r) {
            if (isset($this->fromCustom[$r])) {
                throw new \InvalidArgumentException("alphabet contains duplicate character $r");
            }
            $this->fromCustom[$r] = $i;
        }
    }

    public function encodeBytes(string $data): string
    {
        if ($data === '') {
            return $this->chars[0];
        }
        $len = strlen($data);
        $wrapped = pack('n', $len) . $data;
        $decimal = self::bytesToDecimal($wrapped);
        return $this->intToString($decimal);
    }

    public function decodeBytes(string $s): string
    {
        if ($s === '') {
            throw new \InvalidArgumentException('empty string');
        }
        $decimal = $this->stringToInt($s);
        $raw = self::decimalToBytes($decimal);
        for ($pad = 0; $pad <= 1; $pad++) {
            $buf = str_repeat("\0", $pad) . $raw;
            if (strlen($buf) < 2) {
                continue;
            }
            $unpacked = unpack('n', substr($buf, 0, 2));
            $dataLen = $unpacked[1];
            if (strlen($buf) !== 2 + $dataLen) {
                continue;
            }
            return substr($buf, 2);
        }
        throw new \InvalidArgumentException('invalid encoded data length');
    }

    private function intToString(string $decimal): string
    {
        if (bccomp($decimal, '0') === 0) {
            return $this->chars[0];
        }
        $chars = [];
        $base = (string)$this->base;
        $zero = '0';
        while (bccomp($decimal, $zero) > 0) {
            $rem = (int)bcmod($decimal, $base);
            $chars[] = $this->chars[$rem];
            $decimal = bcdiv($decimal, $base, 0);
        }
        return implode('', array_reverse($chars));
    }

    private function stringToInt(string $s): string
    {
        $runes = preg_split('//u', $s, -1, PREG_SPLIT_NO_EMPTY);
        $n = '0';
        $base = (string)$this->base;
        foreach ($runes as $r) {
            if (!isset($this->fromCustom[$r])) {
                throw new \InvalidArgumentException("invalid character $r");
            }
            $n = bcadd(bcmul($n, $base), (string)$this->fromCustom[$r]);
        }
        return $n;
    }

    private static function bytesToDecimal(string $bytes): string
    {
        $result = '0';
        foreach (str_split($bytes) as $byte) {
            $result = bcadd(bcmul($result, '256'), (string)ord($byte));
        }
        return $result;
    }

    private static function decimalToBytes(string $decimal): string
    {
        if (bccomp($decimal, '0') === 0) {
            return '';
        }
        $bytes = [];
        while (bccomp($decimal, '0') > 0) {
            $rem = (int)bcmod($decimal, '256');
            $bytes[] = chr($rem);
            $decimal = bcdiv($decimal, '256', 0);
        }
        return implode('', array_reverse($bytes));
    }
}

File: java/src/main/java/io/github/vannifan/idmix/RadixCodec.java
package io.github.vannifan.idmix;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;
import java.util.Map;

/** Custom radix text layer codec (default idmix text layer). */
public final class RadixCodec implements ICodec {
    private final int base;
    private final String chars;
    private final Map<Character, Integer> fromCustom;

    public RadixCodec(String alphabet) {
        if (alphabet.length() < 2) {
            throw new IllegalArgumentException("alphabet must have at least 2 unique characters");
        }
        this.base = alphabet.length();
        this.chars = alphabet;
        this.fromCustom = new HashMap<>();
        for (int i = 0; i < alphabet.length(); i++) {
            char c = alphabet.charAt(i);
            if (fromCustom.containsKey(c)) {
                throw new IllegalArgumentException("alphabet contains duplicate character " + c);
            }
            fromCustom.put(c, i);
        }
    }

    public String getAlphabet() { return chars; }
    public int getBase() { return base; }

    @Override
    public String encode(byte[] data) {
        if (data.length == 0) return String.valueOf(chars.charAt(0));
        ByteBuffer buf = ByteBuffer.allocate(2 + data.length);
        buf.order(ByteOrder.BIG_ENDIAN);
        buf.putShort((short) data.length);
        buf.put(data);
        BigInteger n = new BigInteger(1, buf.array());
        return intToString(n);
    }

    @Override
    public byte[] decode(String s) {
        if (s == null || s.isEmpty()) throw new IllegalArgumentException("empty string");
        BigInteger n = stringToInt(s);
        byte[] raw = n.toByteArray();
        if (raw.length > 1 && raw[0] == 0) {
            byte[] trimmed = new byte[raw.length - 1];
            System.arraycopy(raw, 1, trimmed, 0, trimmed.length);
            raw = trimmed;
        }
        for (int pad = 0; pad <= 1; pad++) {
            byte[] buf = new byte[pad + raw.length];
            System.arraycopy(raw, 0, buf, pad, raw.length);
            if (buf.length < 2) continue;
            int dataLen = ((buf[0] & 0xff) << 8) | (buf[1] & 0xff);
            if (buf.length != 2 + dataLen) continue;
            byte[] out = new byte[dataLen];
            System.arraycopy(buf, 2, out, 0, dataLen);
            return out;
        }
        throw new IllegalArgumentException("invalid encoded data length");
    }

    private String intToString(BigInteger n) {
        if (n.signum() == 0) return String.valueOf(chars.charAt(0));
        BigInteger baseBI = BigInteger.valueOf(base);
        StringBuilder sb = new StringBuilder();
        while (n.signum() > 0) {
            BigInteger[] div = n.divideAndRemainder(baseBI);
            sb.append(chars.charAt(div[1].intValue()));
            n = div[0];
        }
        return sb.reverse().toString();
    }

    private BigInteger stringToInt(String s) {
        BigInteger n = BigInteger.ZERO;
        BigInteger baseBI = BigInteger.valueOf(base);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            Integer idx = fromCustom.get(c);
            if (idx == null) throw new IllegalArgumentException("invalid character " + c);
            n = n.multiply(baseBI).add(BigInteger.valueOf(idx));
        }
        return n;
    }
}
